util: Make Parse{Int,UInt}{32,64} use locale independent std::from_chars(…) (C++17) instead of locale dependent strto{l,ll,ul,ull}

practicalswift opened the pull request:
Make `Parse{Int,UInt}{32,64}` use locale independent `std::from_chars(…)` (C++17) instead of locale dependent `strto{l,ll,ul,ull}`.

[About `std::from_chars`](https://en.cppreference.com/w/cpp/utility/from_chars): _"Unlike other parsing functions in C++ and C libraries, `std::from_chars` is locale-independent, non-allocating, and non-throwing."_

laanwj commented:
Concept ACK, see also discussion in https://github.com/bitcoin/bitcoin/pull/20452#issuecomment-731748771

practicalswift commented:
Like #20452 this one is blocked on #20460. That's the reason for the CI failures :)

practicalswift commented:
Could remove "Waiting for author"? I don't know if we have any "Blocked on another PR/issue" (or "Waiting for toolchain upgrade") tag, but this PR is blocked on #20460 :)

jonatack commented:
Concept ACK. Looks very good; hope to see this unblocked with `std::fs`.

practicalswift committed:
util: Introduce ToIntegral<T>(const std::string&) for locale independent parsing using std::from_chars(…) (C++17)

util: Avoid locale dependent functions strtol/strtoll/strtoul/strtoull in ParseInt32/ParseInt64/ParseUInt32/ParseUInt64

fuzz: Assert equivalence between new and old Parse{Int,Uint}{8,32,64} functions

test: Add unit tests for ToIntegral<T>(const std::string&)

laanwj commented:
Concept ACK!

practicalswift commented:
Now that #20460 has been merged I think this PR should be ready for final review :)

laanwj commented:
Nice!
Code review ACK 4747db876154ddd828c03d9eda10ecf8b25d8dc8

MarcoFalke commented on src/util/strencodings.h:
```diff
@@ -94,6 +96,24 @@ constexpr inline bool IsSpace(char c) noexcept {
+    if (first_nonmatching != str.data() + str.size() || error_condition != std::errc{}) {
+        return std::nullopt;
+    }
+    return {result};
```
Can you explain what the `{` are supposed to do? I think they can be removed, as they are confusing as-is. If you want them to check (for whatever reason) that result fits in type `T`, you will have to type `return T{result};`

practicalswift replied:
In this specific case `return result;` should be equivalent to `return {result};` since the return type is `std::optional<T>` and `result` is of type `T`. I can see your point about braces looking a bit odd here: feel free to change :)

Some context behind the "braced return" style: it can help avoid nasty implicit narrowing conversions like in the following example.

```
$ cat without-list-initialization.cpp
#include <cstdint>

struct F {
  F(uint8_t) {}
};

F f() { return 1024; }
$ clang++ -c without-list-initialization.cpp
# No errors
$ echo $?
0
```

```
$ cat with-list-initialization.cpp
#include <cstdint>

struct F {
  F(uint8_t) {}
};

F f() { return {1024}; }
$ clang++ -c with-list-initialization.cpp
with-list-initialization.cpp:7:18: error: constant expression evaluates to 1024 which cannot be narrowed to type 'uint8_t' (aka 'unsigned char') [-Wc++11-narrowing]
$ echo $?
1
```

See item (8) and "narrowing conversions" in https://en.cppreference.com/w/cpp/language/list_initialization :)

practicalswift replied:
FWIW also: https://clang.llvm.org/extra/clang-tidy/checks/modernize-return-braced-init-list.html

MarcoFalke replied:
> FWIW also: https://clang.llvm.org/extra/clang-tidy/checks/modernize-return-braced-init-list.html


`return {baz};` fails when `explicit Foo(Baz)`

maflcko reviewed (commented):
crACK

MarcoFalke commented on src/util/strencodings.h:
```diff
@@ -94,6 +96,24 @@ constexpr inline bool IsSpace(char c) noexcept {
+ *   parsed value is not in the range representable by the type T.
+ */
+template <typename T>
+std::optional<T> ToIntegral(const std::string& str)
```
I am wondering if it makes sense to explain in simple terms the format that is required to be parsed successfully. In regex terms it would be `-?[0-9]+`?

MarcoFalke commented:
Created a follow-up in #23156
